src/libexpr/attr-set.hh

```diff
@@ -7,88 +7,193 @@
 
 namespace nix {
 
-
-class EvalState;
 struct Value;
 
-/* Map one attribute name to its value. */
-struct Attr
-{
-    Symbol name;
-    Value * value;
-    Pos * pos;
-    Attr(Symbol name, Value * value, Pos * pos = &noPos)
-        : name(name), value(value), pos(pos) { };
-    Attr() : pos(&noPos) { };
-    bool operator < (const Attr & a) const
-    {
-        return name < a.name;
-    }
-};
 
-/* Bindings contains all the attributes of an attribute set. It is defined
-   by its size and its capacity, the capacity being the number of Attr
-   elements allocated after this structure, while the size corresponds to
-   the number of elements already inserted in this structure. */
+/* Bindings contains all the attributes of an attribute set. */
 class Bindings
 {
-public:
-    typedef uint32_t size_t;
-
-private:
-    size_t size_, capacity_;
-    Attr attrs[0];
-
-    Bindings(size_t capacity) : size_(0), capacity_(capacity) { }
+    friend class BindingsBuilder;
+    friend size_t valueSize(Value & v);
+
+    class Names {
+        mutable unsigned int * _lexicographicOrder = nullptr; // memoized indexes in names[],values[],positions[] in lexicographic order of names[]
+                                                              // TODO: memoize `attrNames' result (which also can be reused in mapAttrs)
+    public:
+        std::vector<Symbol> symOrder;
+
+        ~Names() {
+            assert(_lexicographicOrder == nullptr); // interned `Names' are never destructed
+        }                                           // `Names' in BindingsBuilder are not complete and should not use `lexicographicOrder()'
+
+        // exclude mutables
+        bool operator==(const Names & that) const { return symOrder == that.symOrder; }
+
+        const unsigned int * lexicographicOrder() const {
+            if (_lexicographicOrder == nullptr) {
+                _lexicographicOrder = new unsigned int[symOrder.size()];
```

Comment:
Shouldn't this be allocated using Boehm?

Reply:
This will never be GC'ed.
Attrset names array are stored in a singleton hash-table (akin to the symbol table) and this ```_lexicographicOrder``` is stored inside the attrset names array to avoid sorting each time when lexicographic order is required.